api/python/provisioner/commands/check.py

```diff
@@ -0,0 +1,578 @@
+
+            res.append(_res)
+
+        return res
+
+    @staticmethod
+    def _communicability(
+            *, args: str,
+            targets: Union[list, tuple, set] = PENDING_SERVERS) -> CheckEntry:
+        """
+        Check if nodes can communicate using `salt test.ping`
+
+        :param args: bmc accessibility check specific parameters and arguments
+        :param targets: target nodes where network checks will be executed
+        :return:
+        """
+        res: CheckEntry = CheckEntry(cfg.Checks.COMMUNICABILITY.value)
+
+        # TODO: need to resolve cfg.ALL_MINIONS alias
+        #  to the list of minions
+        # TODO: create check and return list of nodes which are down
+        if check_salt_minions_are_ready(targets=targets):
+            res.set_passed(target=local_minion_id())
+        else:
+            res.set_fail(target=local_minion_id())
+
+        return res
+
+    @staticmethod
+    def _bmc_accessibility(*, args: str) -> CheckEntry:
+        """
+        Check BMC accessibility
+
+        :param args: bmc accessibility check specific parameters and arguments
+        :return:
+        """
+        _BMC_CHECK_CMD = 'ipmitool chassis status | grep "System Power"'
+
+        res: CheckEntry = CheckEntry(cfg.Checks.BMC_ACCESSIBILITY.value)
+
+        minion_id = local_minion_id()
+
+        try:
+            _res = cmd_run(_BMC_CHECK_CMD, targets=minion_id,
+                           fun_kwargs=dict(python_shell=True))
+            # TODO: parse command output if necessary
+            res.set_passed(target=minion_id, comment=str(_res))
+        except SaltCmdResultError as e:
+            res.set_fail(target=minion_id, comment=str(e.reason))
+
+        return res
+
+    @staticmethod
+    def _connectivity(*, servers: Union[list, tuple, set] = PENDING_SERVERS,
+                      args: str) -> CheckEntry:
+        """
+        Check servers connectivity. Ping servers to check for availability
+
+        :param Union[list, tuple, set] servers: servers for pinging
+        :param str args: possible arguments for connectivity checks
+        :return:
+        """
+        # ping command template
+        # -c 1 means count of sending ECHO_REQUEST packets
+        # -W 1 means timeout to wait for a response
+        _PING_CMD_TMPL = "ping -c 1 -W 1 {server_addr}"
+
+        res: CheckEntry = CheckEntry(cfg.Checks.CONNECTIVITY.value)
+
+        for addr in servers:
+            # NOTE: check ping of 'srvnode-2' from 'srvnode-1'
+            # and vise versa
+            # TODO: which targets do we need to use? Because we need to
+            #  check cross connectivity if nodes > 2 and all possible pairs
+            targets = set(servers) - {addr}
+            targets = (cfg.LOCAL_MINION if not targets
+                       else next(iter(targets)))  # takes just one node
+
+            cmd = _PING_CMD_TMPL.format(server_addr=addr)
+            try:
+                cmd_run(cmd, targets=targets)
```

Comment:
here `targets` are nodes on which ping is run, and `addr` is the node which is pinged. So, is it ok to set `target` to that value?

Reply:
Yes, we are trying to check cross-connectivity between nodes. Also in future we need to improve this to check all possible combinations of pairs (target host which performs ping command and destination host). The code already had the necessary TODO comments:
```python
            # TODO: which targets do we need to use? Because we need to
            #  check cross connectivity if nodes > 2 and all possible pairs
            targets = set(servers) - {addr}
            targets = (cfg.LOCAL_MINION if not targets
                       else next(iter(targets)))  # takes just one node
```

Reply:
the check fails if any of source services fails to connect to destination one.
Let's imaging that:
- we have 5 nodes
- node 1 and node 2 are not ping-able from some of other ones.

What would be passed to `set_fail`? My guess:

- set_fail(target={node2, node3, node4, node5})
- set_fail(target={node1, node2, node4, node5})

Don't you think that:
- set_fail(target=node1)
- set_fail(target=node2)

would be clearer and easier to use if we want to extract a node that fails the check?

Reply:
> * set_fail(checked_target=node1)
> * set_fail(checked_target=node2)

Yes, this method will be used to create a validation result. At the current moment, we take just one node for validation. During our improvement, we will have as many records as the number of pairs. I used this approach in other checks such as `_logs_are_good`, `_network`. It is the most simple way to do such validation instead of the creation of some new check instance which encapsulates a series of check results.

Reply:
So this method will return list of `CheckEntry` instances.